plonky2x/src/frontend/builder/mod.rs

```diff
@@ -147,18 +166,106 @@ impl<L: PlonkParameters<D>, const D: usize> CircuitBuilder<L, D> {
         };
 
         let data = self.api.build();
-        CircuitBuild { data, io: self.io }
+
+        let mut async_hint_indices = Vec::new();
+
+        for (i, generator) in data.prover_only.generators.iter().enumerate() {
+            if generator.0.id().starts_with("--async") {
+                async_hint_indices.push(i);
+            }
+        }
+
+        assert_eq!(async_hint_indices.len(), self.async_hints.len());
+
+        let mut async_hints = BTreeMap::new();
+        for (key, gen) in async_hint_indices.iter().zip(self.async_hints) {
+            async_hints.insert(*key, gen);
+        }
+
+        CircuitBuild {
+            data,
+            io: self.io,
+            async_hints,
+        }
     }
 
     pub fn mock_build(mut self) -> MockCircuitBuild<L, D> {
         if !self.sha256_requests.is_empty() {
             self.curta_constrain_sha256();
         }
-        let mock_circuit = self.api.mock_build();
+
+        for (index, gen_ref) in self
```

Comment:
My feeling here is instead of duplicating all this code between `mock_build` and `build`, we should have a "pre_build` hook that contains all this code and then just call the hook in both `mock_build` and `build`. What do you think?

Reply:
Yeah it would be nicer, I will add it